cmd/snap/cmd_reboot.go

```diff
@@ -0,0 +1,110 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"fmt"
+
+	"github.com/jessevdk/go-flags"
+
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/i18n"
+)
+
+type cmdReboot struct {
+	clientMixin
+	Positional struct {
+		Label string
+	} `positional-args:"true" required:"true"`
+
+	RunMode     bool `long:"run"`
+	InstallMode bool `long:"install"`
+	RecoverMode bool `long:"recover"`
+}
+
+var shortRebootHelp = i18n.G("Reboot into select recovery mode")
+var longRebootHelp = i18n.G(`
+The reboot command reboots the system into a special mode of the selected
+recovery system.
+`)
+
+func init() {
+	addCommand("reboot", shortRebootHelp, longRebootHelp, func() flags.Commander {
+		return &cmdReboot{}
+	}, map[string]string{
+		// TRANSLATORS: This should not start with a lowercase letter.
+		"run": i18n.G("Boot into run mode"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		"install": i18n.G("Boot into install mode"),
+		// TRANSLATORS: This should not start with a lowercase letter.
+		"recover": i18n.G("Boot into recover mode"),
+	}, []argDesc{
+		{
+			// TRANSLATORS: This needs to begin with < and end with >
+			name: i18n.G("<label>"),
+			// TRANSLATORS: This should not start with a lowercase letter.
+			desc: i18n.G("The recovery system label"),
+		},
+	})
+}
+
+func (x *cmdReboot) modeFromCommandline() (*client.SystemAction, error) {
+	var action client.SystemAction
+	for _, arg := range []struct {
+		enabled bool
+		mode    string
+	}{
+		{x.RunMode, "run"},
+		{x.RecoverMode, "recover"},
+		{x.InstallMode, "install"},
+	} {
+		if !arg.enabled {
+			continue
+		}
+		if action.Mode != "" {
+			return nil, fmt.Errorf("Please specify a single mode")
+		}
+		action.Mode = arg.mode
+	}
+	// XXX: should we have a default here?
+	if action.Mode == "" {
+		return nil, fmt.Errorf("Please specify a mode, see --help")
+	}
+
+	return &action, nil
+}
+
+func (x *cmdReboot) Execute(args []string) error {
+	if len(args) > 0 {
+		return ErrExtraArgs
+	}
+
+	action, err := x.modeFromCommandline()
+	if err != nil {
+		return err
+	}
+
+	if err := x.client.DoSystemAction(x.Positional.Label, action); err != nil {
+		return fmt.Errorf("cannot reboot: %v", err)
+	}
+
+	fmt.Fprintf(Stdout, "Reboot into %q with mode %q scheduled.\n", x.Positional.Label, action.Mode)
```

Comment:
Stderr?

Reply:
why stderr ?

Reply:
It feels like a log message rather than the actual output of the command.

Reply:
Any suggestions for a better message? I agree it sounds very technical (and bit log like). Maybe some native speaker like @anonymouse64 has an idea?

Reply:
/me puts on bikeshed hat

how about `"Rebooting into %q %q mode"` that would look like:
```
$ snap reboot 20201129 --recover
Rebooting into 20201129 recovery mode
$ snap reboot 20201129 --run
Rebooting into 20201129 run mode
$ snap reboot 20201129 --install
Rebooting into 20201129 install mode
```

I think the word "scheduled" is a little misleading because yes it is technically scheduled, but it's scheduled for effectively very very soon, so to say it's scheduled makes folks think it will happen a bit later which it won't. Especially for a command that says "reboot", I think it should use the present tense of the verb reboot since that's what it's doing, it didn't already do the reboot, it's actively going to do it right now.

Reply:
the suggestion seems fine to me

Reply:
I updated it now to use: "Rebooting into %q %q mode" as suggested.